File: files/CH-194.R
library(tidyverse)
library(readxl)

path = "files/CH-194 Pattern Length.xlsx"
input = read_excel(path, range = "B2:C7")
test  = read_excel(path, range = "G2:H7")

result <- input %>%
  separate_rows(Pattern, sep = " ") %>%
  mutate(cn = consecutive_id(Pattern), .by = Date) %>%
  mutate(n = n(), .by = c(Date, cn)) %>%
  filter(n == max(n), .by = Date) %>%
  unite("Length", Pattern, n, sep = "") %>%
  select(-cn) %>%
  distinct()

all.equal(result, test)
#> [1] TRUE

File: files/CH-194.py
import pandas as pd

path = "CH-194 Pattern Length.xlsx"
input = pd.read_excel(path, usecols="A:C", skiprows=1, nrows=5)
test = pd.read_excel(path, usecols="G:H", skiprows=1, nrows=5, dtype={'Length': str}).rename(columns=lambda x: x.split('.')[0])

input = input.assign(Pattern=input['Pattern'].str.split(" ")).explode('Pattern').reset_index(drop=True)
input['cn'] = (input.groupby('Date')['Pattern'].transform(lambda x: (x != x.shift()).cumsum()))
input['n'] = input.groupby(['Date', 'cn'])['Pattern'].transform('size')
max_df = input.loc[input.groupby('Date')['n'].idxmax()].copy()
max_df['Length'] = max_df['Pattern'] + max_df['n'].astype(str)
result = max_df[["Date", "Length"]].reset_index(drop=True)

print(result.equals(test)) # True